Add history of actions in designer and ability to undo/redo actions

jasonchung1871 opened the pull request:
# Description

When the form is mounted or when a form is fetched, we cache the original schema. When performing an action, we add that action to history of actions in the FormDesigner. These actions include importing a schema, adding, removing, moving, and updating a component. Added an undo button and redo button to the FormDesigner toolbar, both buttons have a counter displaying how many times the user can perform an undo or redo operation. When pressing undo,  we go back one action and generate a form schema from the original schema with all patches up until the previous action and set the form schema to the generated one. When pressing redo,  we go forward one action and generate a form schema from the original schema with all patches up until the next action and set the form schema to the generated one. When adding an action to the history, if we're not at the end of the history, we remove all actions after the current action.

## Types of changes

New feature (non-breaking change which adds functionality)
- Undo button, reverts to previous action in history
- Redo button, reverts to next action in history

## Checklist

<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

Something that I could do is add tests to prove that the feature works.

bcgov-nr-csst commented:
PR Deployed at: https://chefs-dev.apps.silver.devops.gov.bc.ca/pr-283

jujaga commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -369,15 +443,135 @@ export default {
-      this.setDirtyFlag(false);
     },
-    onChangeMethod() {
+    onChangeMethod(_changed, _flags, _modified) {
```
`onChangeMethod` is a bit ambiguous as to what event it is handling, and for what object/context. I would propose creating another method called something like `onSchemaChange` as it makes it more clear the intent of this method (do something with the schema when it changes), and  then calling that function within the `onChangeMethod` function. IIRC `onChangeMethod` is supposed to mirror/override the internal form.io event hooks, so it would be safer to functionalize out our history changes into its own invokable method instead.

jujaga commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -369,15 +443,135 @@ export default {
+      // or redo button
+      if (!this.patchHistoryUndo && !this.patchHistoryRedo) {
+        // flags and modified are defined when a component is added
+        if (_flags !== undefined && _modified !== undefined) {
+          if (this.componentAdded !== null) {
+            this.addPatchToHistory();
+          } else {
+            this.resetHistoryFlags();
+          }
+        } else {
+          if (this.componentRemoved) {
+            // Component was removed..
+            this.addPatchToHistory();
+          } else if (this.componentMoved) {
+            this.addPatchToHistory();
+          } else {
+            this.resetHistoryFlags();
+          }
+        }
```
- Double check the new variables such as componentAdded - do we need to be initalizing it as null, or can we keep it as just a simpler boolean? If we can do the latter, it can simplify the logic needed.
- This entire section of logic seems like it doesn't need to be branched because in most situations, you're either just calling `addPatchToHistory` or `resetHistoryFlags` methods when something changes (or not). You could potentially simplify this entire block to the following, ignoring what kind of change it was, and ignoring the `_flags` and `_modified` too (pseudocode... don't know if there's other side effects to worry about?):

```js
if (this.componentAdded || this.componentMoved || this.componentRemoved) {
  this.addPatchToHistory();
} else {
  this.resetHistoryFlags();
}
```

jasonchung1871 replied:
first off, sorry for the wall of text, but this is for anyone else that may run into a similar issue!

the first one was a mistake on my part, initially i was storing the component in the componentAdded variable so that i could store the new component in the history after the user clicks "save". i've reverted it back to a boolean and now it is much cleaner, simpler, and unpacks/repacks correctly since we're using fast-json-patch.

the reason why it isn't simplified is because i wasn't sure at first how complex it would become. formio calls the change method for absolutely every change even if the user doesn't click save or if they click cancel/remove. we unfortunately need the `flags` and `modified` parameters to verify that the user clicked save and actually added the component to the form. i also had to create my own flags for when we started an action such as add, remove, move, and update. those flags are set to true when the actions begin in their respective functions `onAddComponentMethod`, `onRemoveComponentMethod`, `onUpdateComponentMethod` and then set to false once an action is completed.

for example, the save process alone looks like:
drag and drop -> `onUpdateComponentMethod()` -> `onChangeMethod(changed, undefined, undefined)` -> `onAddComponent()`-> `onUpdateComponentMethod()` -> `onChangeMethod(changed, undefined, undefined)`
save -> `onChangeMethod(changed, flags, modified)` -> `onChangeMethod(changed, undefined, undefined)`
canceled after drag and drop -> `onRemoveComponentMethod()` -> `onChangeMethod(changed, undefined, undefined)`
removed after drag and drop -> `onRemoveComponentMethod()` -> `onChangeMethod(changed, undefined, undefined)`

after writing all this, i realized we don't really need fine-grained control over the actions since we're not writing the json comparator/patcher. i've removed the `onUpdateComponentMethod` method since an update is similar to an "add" with the `flags` and `modified` parameters defined. if we did want fine-grained control, then we could obtain the information for exactly what is being added/updated/removed and know if the action was saved or canceled/removed.

jujaga commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -369,15 +443,135 @@ export default {
+    },
+    getPatch(_idx) {
+      // Generate the form from the original schema
+      var form = JSON.parse(JSON.stringify(this.$options.originalSchema));
```
- Do not use var. Declare with either const or let only.
- If I understand this correctly, you can use deepClone to achieve the same thing instead of stringify/parse chaining it.

jujaga commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -369,15 +443,135 @@ export default {
+
+      // Get the differences between the last patch 
+      // and the current form
+      var form = this.getPatch(++this.patchIndex);
+      var patch = compare(form, this.formSchema);
```
Do not use var. Declare with either const or let only.

jujaga commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -369,15 +443,135 @@ export default {
+      var form = JSON.parse(JSON.stringify(this.$options.originalSchema));
+      if (this.patchIndex > -1 && this.patchHistory.length > 0) {
+        // Apply all patches until we reach the requested patch
+        for (var i = -1; i < _idx; i++) {
```
Do not use var. Declare with either const or let only.

jujaga commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -369,15 +443,135 @@ export default {
+
+      this.resetHistoryFlags();
+    },
+    getPatch(_idx) {
```
`_idx` shouldn't be underscored as it is actually being used in the body, that should not be prefixed with an underscore as that is contrary to the purposefully "unused" argument variable declaration. It should just be `idx` instead.

jujaga commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -369,15 +443,135 @@ export default {
+      }
+    },
+    canUndoPatch() {
+      if (this.patchHistory.length == 0) return false;
+      if (this.patchIndex > this.patchHistory.length) return false;
+      if (this.patchIndex < 0) return false;
+      return true;
```
```js
return this.patchHistory.length && this.patchIndex >= 0 && this.patchIndex < this.patchHistory.length;
```

jujaga commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -334,6 +399,15 @@ export default {
+          this.addPatchToHistory();
+          this.patchHistoryUndo = false;
+          this.patchHistoryRedo = false;
+          this.componentAdded = false;
+          this.componentRemoved = false;
+          this.componentUpdated = false;
+          this.componentMoved = false;
```
You already have a `resetHistoryFlags` method defined. You should be using reusing that.

```js
this.resetHistoryFlags();
```

jujaga commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -413,6 +607,12 @@ export default {
         this.saving = false;
       }
     },
+    async onUndoClick() {
+      this.undoPatchFromHistory();
+    },
+    async onRedoClick() {
+      this.redoPatchFromHistory();
+    },
```
Not all methods need to be declared as async functions. For these ones, you can make these just normal functions. In general, you only want to make these methods async if and only if:

1. The method body has at least one `await`
2. The return result is expected to be of Promise type.
3. You know the methods invoked are asynchronous in nature and cannot behave synchronously.

Double check the rest of your new methods as well.

jujaga commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -477,11 +677,18 @@ export default {
   },
+  mounted() {
+    if (!this.formId) {
+      // We are creating a new form, so we obtain the original schema here.
+      this.$options.originalSchema = JSON.parse(JSON.stringify(deepFreeze(deepClone(this.formSchema))));
+      deepFreeze(this.$options.originalSchema);
```
If we are already deepCloning an object, there is no need to do a JSON stringify/parse chain as it is already doing the exact same thing. If I understand the intent of code, this can be a one-liner:

```js
this.$options.originalSchema = deepFreeze(deepClone(this.formSchema));
```

jasonchung1871 replied:
Ah, the reason for this was because something kept making the originalSchema observable and changing it even though it should be read only. I think I found that it was occurring when generating the final schema in the getPatch method.

jujaga reviewed (changes requested):
This is quite an interesting feature addition. Fantastic work! 👍 
That being said, there are a good handful of code style and considerations I'd like for you to take a look at. Please take a look at the detailed comments below and address them when you get a chance.

jasonchung1871 commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -176,6 +216,7 @@ export default {
     },
     versionId: String,
   },
+  originalSchema: null,
```
i could not for the life of me figure out why the originalSchema variable was being modified even though it was made to be read-only with deepFreeze, i've figured that out now though. i was trying out many different things to try and just prevent that variable from being modified after loading the schema.

jasonchung1871 commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -162,6 +199,9 @@ import { formService } from '@/services';
 import { IdentityMode, NotificationTypes } from '@/utils/constants';
 import { generateIdps } from '@/utils/transformUtils';
 
+var deepFreeze = require('deep-freeze');
+import { compare, applyPatch, deepClone } from 'fast-json-patch';
```
the last 3 imports are not mine, i just went with what was already there. i'll change them though

jasonchung1871 commented on app/frontend/src/components/designer/FormDesigner.vue:
```diff
@@ -317,6 +376,12 @@ export default {
+        if (this.patchHistory.length === 0) {
+          // We are fetching an existing form, so we get the original schema here because
+          // using the original schema in the mount will give you the default schema
+          this.$options.originalSchema = JSON.parse(JSON.stringify(deepFreeze(deepClone(this.formSchema))));
+          deepFreeze(this.$options.originalSchema);
```
i've figured out what area was causing the original schema to be modified and have removed the deepFreeze as well!

TimCsaky commented:
just tried in the browser..  looking good!
minor point, the 'redo' button is active after adding the first component to a blank form. might have something to do with the default data values?

jasonchung1871 commented:
> just tried in the browser.. looking good! minor point, the 'redo' button is active after adding the first component to a blank form. possibly some incrementing issues. might have something to do with the default data values?

thank you! this was an issue on my part, i removed the increment to the history index when testing the limit on max actions in the history. i've added the increment back in, and all is well!
